Model builder Install How-To and Price Prediction Tutorial (#12614)
dotnet/docs: docs/machine-learning/tutorials/predict-prices-with-model-builder.md

---
title: 'Predict prices using regression with Model Builder'
description: This tutorial illustrates how to build a regression model using ML.NET Model Builder to predict prices, specifically, New York City taxi fares.
author: luisquintanilla
ms.author: luquinta
ms.date: 05/31/2019
ms.topic: tutorial
ms.custom: mvc
#Customer intent: As a non-developer, I want to use Model Builder to automatically generate a model to predict prices using Model Builder. 
---

# Predict prices using regression with Model Builder

This tutorial illustrates how to build a [regression model](../resources/glossary.md#regression) using ML.NET Model Builder to predict prices, specifically, New York City taxi fares. The regression machine learning task can be utilized in various scenarios where the target or label to predict is numerical.

In this tutorial, you learn how to:
> [!div class="checklist"]
> * Prepare and understand the data
> * Choose a scenario
> * Load the data
> * Train the model
> * Evaluate the model
> * Use the model for predictions

> ## Pre-requisites

Reviewer: We probably should add the Model Builder is in preview call out here

For a list of pre-requisites and installation instructions, visit the [Model Builder installation guide](../how-to-guides/install-model-builder.md).

## Create a console application

1. Create a **.NET Core Console Application** called "TaxiFarePrediction".

1. Create a directory named *Data* in your project to store the data set files.

1. Install the **Microsoft.ML** NuGet Package:

    In **Solution Explorer**, right-click the *TaxiFarePrediction* project and select **Manage NuGet Packages**. Choose "nuget.org" as the Package source, select the **Browse** tab, search for **Microsoft.ML**, select the package in the list, and select the **Install** button. Select the **OK** button on the **Preview Changes** dialog and then select the **I Accept** button on the **License Acceptance** dialog if you agree with the license terms for the packages listed.

> ## Prepare and Understand the Data

Reviewer: Lowercase `understand` and `data`

> 1. Download the [taxi-fare-train.csv](https://github.com/dotnet/machinelearning/blob/master/test/data/taxi-fare-train.csv) and data set and save it to the *Data* folder you've created at the previous step. THis data set will be used to train and evaluate the machine learning model. This data sets is originally from the [NYC TLC Taxi Trip data set](http://www.nyc.gov/html/tlc/html/about/trip_record_data.shtml).

Reviewer: Is `and data set` a typo?

> 1. Download the [taxi-fare-train.csv](https://github.com/dotnet/machinelearning/blob/master/test/data/taxi-fare-train.csv) and data set and save it to the *Data* folder you've created at the previous step. THis data set will be used to train and evaluate the machine learning model. This data sets is originally from the [NYC TLC Taxi Trip data set](http://www.nyc.gov/html/tlc/html/about/trip_record_data.shtml).

Reviewer: THis --> this

> 1. Download the [taxi-fare-train.csv](https://github.com/dotnet/machinelearning/blob/master/test/data/taxi-fare-train.csv) and data set and save it to the *Data* folder you've created at the previous step. THis data set will be used to train and evaluate the machine learning model. This data sets is originally from the [NYC TLC Taxi Trip data set](http://www.nyc.gov/html/tlc/html/about/trip_record_data.shtml).

Reviewer: will be --> is

> 1. Download the [taxi-fare-train.csv](https://github.com/dotnet/machinelearning/blob/master/test/data/taxi-fare-train.csv) and data set and save it to the *Data* folder you've created at the previous step. THis data set will be used to train and evaluate the machine learning model. This data sets is originally from the [NYC TLC Taxi Trip data set](http://www.nyc.gov/html/tlc/html/about/trip_record_data.shtml).

Reviewer: This data sets --> data set

1. In **Solution Explorer**, right-click the *taxi-fare-train.csv* file and select **Properties**. Under **Advanced**, change the value of **Copy to Output Directory** to **Copy if newer**.

> 1. Open the **taxi-fare-train.csv** data set and look at column headers in the first row. Take a look at each of the columns. Understand the data and decide which columns are **features** and which one is the **label**.

Reviewer: Not sure this is entirely clear. How do they understand which ones should be used as features? Can we elaborate on this?

Author: #resolved

> The `label` is the column you want to predict. The identified `Features`are the inputs you give the model to predict the `Label`.

Reviewer: Consistent capitalization of `label` and `features`

        

        // 3. Use PredictionEngine to use model on input data
        ModelOutput result = predictionEngine.Predict(input);

        // 4. Return prediction result
        return result;
    }
    ```

7. Add the following code to the `Main` method and run the application.

    ```csharp
    // Create sample data
    ModelInput input = new ModelInput()
    {
        Vendor_id = "CMT",
        Rate_code = 1,
        Passenger_count = 1,
        Trip_time_in_secs = 1271,
        Trip_distance = 3.8f,
        Payment_type = "CRD"
    };

    // Make prediction
    ModelOutput prediction = ConsumeModel(input);

    // Print Prediction
    Console.WriteLine($"Predicted Fare: {prediction.Score}");
    Console.ReadKey();
    ```

    The output generated by the program should look similar to the snippet below:

    ```bash
    Predicted Fare: 16.82245
    ```

## Next Steps

In this tutorial, you learned how to:
> [!div class="checklist"]
> * Prepare and understand the data
> * Choose a scenario
> * Load the data
> * Train the model
> * Evaluate the model
> * Use the model for predictions

> - [Deploy a model to Azure Functions](https://docs.microsoft.com/en-us/dotnet/machine-learning/how-to-guides/serve-model-serverless-azure-functions-ml-net)

Reviewer: Should this be here?

Author: Maybe. In those how-to's we just require that the user has a pre-trained model which is what they'll be building with this tutorial.

Reviewer: Suggest making it consistent with the other tutorials i.e. In this tutorial, you *learned* how to
- blah

Then have a sentence about what to do next
- with a button linking to the next thing

See attached screenshotScreen Shot 2019-06-10 at 4.57.17 PM

Author: #resolved

author: luisquintanilla
ms.author: luquinta
ms.date: 06/26/2019
ms.topic: tutorial
ms.custom: mvc
#Customer intent: As a non-developer, I want to use Model Builder to automatically generate a model to predict prices using Model Builder. 
---

# Predict prices using regression with Model Builder

This tutorial illustrates how to build a [regression model](../resources/glossary.md#regression) using ML.NET Model Builder to predict prices, specifically, New York City taxi fares. The regression machine learning task can be utilized in scenarios where the target or label to predict is numerical.

In this tutorial, you learn how to:
> [!div class="checklist"]
> * Prepare and understand the data
> * Choose a scenario
> * Load the data
> * Train the model
> * Evaluate the model
> * Use the model for predictions

> [!NOTE]
> Model Builder is currently in Preview.

## Pre-requisites

For a list of pre-requisites and installation instructions, visit the [Model Builder installation guide](../how-to-guides/install-model-builder.md).

## Create a console application

1. Create a **.NET Core Console Application** called "TaxiFarePrediction".

> 1. Create a directory named *Data* in your project to store the data set files.

Reviewer: Would the steps be clearer if we created the Data directory at the same time as we download the data?

Author: #resolved

    In **Solution Explorer**, right-click the *TaxiFarePrediction* project and select **Manage NuGet Packages**. Choose "nuget.org" as the Package source, select the **Browse** tab, search for **Microsoft.ML**, select the package in the list, and select the **Install** button. Select the **OK** button on the **Preview Changes** dialog and then select the **I Accept** button on the **License Acceptance** dialog if you agree with the license terms for the packages listed.

## Prepare and understand the data

1. Download the [taxi-fare-train.csv](https://github.com/dotnet/machinelearning/blob/master/test/data/taxi-fare-train.csv) data set and save it to the *Data* folder you've created at the previous step. This data set is used to train and evaluate the machine learning model. This data set is originally from the [NYC TLC Taxi Trip data set](http://www.nyc.gov/html/tlc/html/about/trip_record_data.shtml).

1. In **Solution Explorer**, right-click the *taxi-fare-train.csv* file and select **Properties**. Under **Advanced**, change the value of **Copy to Output Directory** to **Copy if newer**.

Each row in the `taxi-fare-train.csv` data set contains details of trips made by a taxi. 

> 1. Open the **taxi-fare-train.csv** data set and look at column headers in the first row. Take a look at each of the columns. 

Reviewer: ```suggestion
1. Open the **taxi-fare-train.csv** data set
```

> The provided data set contains the following columns:

Reviewer: ```suggestion
    The provided data set contains the following columns:
```

> * **vendor_id:** The ID of the taxi vendor is a feature.

Reviewer: I think we can indent this whole block into the numbered section. This would improve readability

Author: #resolved

To train your model, you need to select from the list of available machine learning scenarios provided by Model Builder. In this case, the scenario is `Price Prediction`. For a more extensive list visit the [Model Builder overview article](../automate-training-with-model-builder?branch=pr-en-us-12557#scenarios).

1. In **Solution Explorer**, right-click the *TaxiFarePrediction* project, and select **Add** > **Machine Learning**.
1. In the scenario step of the Model Builder tool, select *Price Prediction* scenario.

## Load the data

Model Builder accepts data from two sources, a SQL Server database or a local file csv or tsv file. For more information on data format requirements, visit the following [link](../how-to-guides/install-model-builder?branch=pr-en-us-12557#limitations).

1. In the data step of the Model Builder tool, select *File* from the data source dropdown.
1. Select the button next to the *Select a file* text box and use File Explorer to browse and select the *taxi-fare-test.csv* in the *Data* directory
1. Choose *fare_amount* in the *Label or Column to Predict* dropdown and navigate to the train step of the Model Builder tool.

## Train the model

> The machine learning task used to train the price prediction model in this tutorial is regression. ML.NET supports various regression algorithms. During the model training process, Model Builder will train separate models using different regression algorithms and hyper-parameters to try and train the best performing model for your dataset.

Reviewer: ```suggestion
The machine learning task used to train the price prediction model in this tutorial is regression. During the model training process, Model Builder trains separate models using different regression algorithms and settings to find the best performing model for your dataset.
```

- Best accuracy displays the accuracy of the best performing model found by Model Builder so far. Higher accuracy means the model predicted more correctly on test data. 
- Best algorithm displays the name of the best performing algorithm performed found by Model Builder so far.
- Last algorithm displays the name of the algorithm most recently used by Model Builder to train the model.

Once training is complete, navigate to the evaluate step.

## Evaluate the model

The result of the training step will be one model which had the best performance. In the evaluate step of the Model Builder tool, the output section, will contain the algorithm used by the best performing model in the *Best Model* entry along with metrics in *Best Model Quality (RSquared)*. Additionally, a summary table containing top five models and their metrics. Visit the following link to learn more about [evaluating model metrics](https://docs.microsoft.com/dotnet/machine-learning/resources/metrics).

If you're not satisfied with your accuracy metrics, some easy ways to try and improve model accuracy are to increase the amount of time to train the model or use more data.

## Use the model for predictions

> Two projects will be created in `C:\Users\%USERNAME%\AppData\Local\Temp\MLVSTools` directory.

Reviewer: Does the user need to know that this is where the projects are? Couldn't this be transparent?

Author: They don't. Most likely the user will just add the project to the current solution immediately after training. In the event they don't, they have the option to add the generated projects at their leisure by referencing that directory.

Author: #resolved
